File: src/python/linked_list/swap_nodes_in_pairs.py
from python.tools.list_node import ListNode


def swapPairs(self, head: ListNode) -> ListNode:
    """
    88ms 7.43%
    13MB 89.91%
    和32秒那个最快的思路是一样的......
    :param self:
    :param head:
    :return:
    """
    cur = ListNode(0)
    cur.next = head
    head = cur

    while cur and cur.next and cur.next.next:
        temp = cur.next.next
        cur.next.next = temp.next
        temp.next = cur.next
        cur.next = temp
        cur = cur.next.next

    return head.next


def swapPairs2(self, head: ListNode) -> ListNode:
    """
    24ms
    """
    if not head or not head.next:
        return head
        
    tmp = head.next
    head.next = self.swapPairs(head.next.next)
    tmp.next = head

    return tmp

File: leetcode-java/src/classify/recursion/SwapNodesInPairs.java
package classify.recursion;

import tools.ListNode;

/**
 * @author yutiantang
 * @create 2021/4/6 22:36
 */
public class SwapNodesInPairs {

    /**
     * 0ms      100.00%
     * 36.1MB   37.73%
     * @param head
     * @return
     */
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode second = head.next;
        ListNode third = head.next.next;
        head.next = swapPairs(third);
        second.next = head;

        return second;
    }

    /**
     * 0ms      100.00%
     * 39.2MB   8.37%
     * @param  head 1->2->3->4->5
     * @return ListNode 2->1->4->3->5
     */
    public ListNode swapPairs2(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode cur = head, tempLeft = cur, tempRight;
        head = head.next;

        while (cur != null && cur.next != null) {
            tempRight = cur.next.next;

            tempLeft.next = cur.next;
            cur.next.next = cur;

            cur.next = tempRight;

            tempLeft = cur;
            cur = cur.next;
        }

        return head;
    }
}
